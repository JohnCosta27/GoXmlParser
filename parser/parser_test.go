package parser

import (
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"johncosta.tech/xmlparse/AST"
	"johncosta.tech/xmlparse/lexer"
)

// ------------- INDIVIDUAL PARSES ------------- //

func TestSimple(t *testing.T) {
  ast, err := Parse(lexer.Tokenize("<a>hello</a>"))

  if (err != nil) {
    t.Error(err)
    t.FailNow()
  }

  snaps.MatchSnapshot(t, ast.Print(0))
}

func TestNesting(t *testing.T) {
  ast, err := Parse(lexer.Tokenize("<a><b>Something</b></a>"))

  if (err != nil) {
    t.Error(err)
    t.FailNow()
  }

  snaps.MatchSnapshot(t, ast.Print(0))
}

func TestAttributes(t *testing.T) {
  ast, err := Parse(lexer.Tokenize(`<a hello="world"></a>`))

  if (err != nil) {
    t.Error(err)
    t.FailNow()
  }

  snaps.MatchSnapshot(t, ast.Print(0))
}

func TestSelfClosing(t *testing.T) {
  ast, err := Parse(lexer.Tokenize(`<atag bruh="hello-world" />`))

  if (err != nil) {
    t.Error(err)
    t.FailNow()
  }

  snaps.MatchSnapshot(t, ast.Print(0))
}

func TestSibling(t *testing.T) {
  ast, err := Parse(lexer.Tokenize("<a><b>Hello</b><c>World</c></a>"))

  if (err != nil) {
    t.Error(err)
    t.FailNow()
  }

  if (ast.ElementSuffix.Content.Type != AST.CONTENT_ELEMENT) {
    t.Log("Expected content to be element")
    t.FailNow()
  }

  el := ast.ElementSuffix.Content.Element
  if (el.OpenTag.NAME.TokenContent != "b") {
    t.Log("Expected first sibling to be `b`")
    t.FailNow()
  }

  otherEl := ast.ElementSuffix.Content.Content
  if (otherEl.Type != AST.CONTENT_ELEMENT) {
    t.Log("Expected content to be element")
    t.FailNow()
  }

  if (otherEl.Element.OpenTag.NAME.TokenContent != "c") {
    t.Log("Expected first sibling to be `c`")
    t.FailNow()
  }

}

func TestMedium(t *testing.T) {
  ast, err := Parse(lexer.Tokenize("<a><b>hello</b><c><d>world</d><e>world2</e><f>world3</f></c></a>"))

  if err != nil {
    t.Error(err)
    t.FailNow()
  }

  snaps.MatchSnapshot(t, ast.Print(0))
}

func TestBigObject(t *testing.T) {
  ast, err := Parse(lexer.Tokenize(`<Tests xmlns="http://www.adatum.com">
  <Test TestId="0001" TestType="CMD">
    <Name>Convert number to string</Name>
    <CommandLine>Examp1.EXE</CommandLine>
    <Input>1</Input>
    <Output>One</Output>
  </Test>
  <Test TestId="0002" TestType="CMD">
    <Name>Find succeeding characters</Name>
    <CommandLine>Examp2.EXE</CommandLine>
    <Input>abc</Input>
    <Output>def</Output>
  </Test>
  <Test TestId="0003" TestType="GUI">
    <Name>Convert multiple numbers to strings</Name>
    <CommandLine>Examp2.EXE /Verbose</CommandLine>
    <Input>123</Input>
    <Output>One Two Three</Output>
  </Test>
  <Test TestId="0004" TestType="GUI">
    <Name>Find correlated key</Name>
    <CommandLine>Examp3.EXE</CommandLine>
    <Input>a1</Input>
    <Output>b1</Output>
  </Test>
  <Test TestId="0005" TestType="GUI">
    <Name>Count characters</Name>
    <CommandLine>FinalExamp.EXE</CommandLine>
    <Input>This is a test</Input>
    <Output>14</Output>
  </Test>
  <Test TestId="0006" TestType="GUI">
    <Name>Another Test</Name>
    <CommandLine>Examp2.EXE</CommandLine>
    <Input>Test Input</Input>
    <Output>10</Output>
  </Test>
</Tests>`))

  if (err != nil) {
    t.Error(err)
    t.FailNow()
  }

  snaps.MatchSnapshot(t, ast.Print(0))
}

func TestDoubleTesting(t *testing.T) {
  ast, err := Parse(lexer.Tokenize(`
    <a>
      <c>Hello</c>
      <b>
        <d>123</d>
        <d>321</d>
      </b>
    </a>
  `))

  if err != nil {
    t.Error(err)
    t.FailNow()
  }

  snaps.MatchSnapshot(t, ast.Print(0))
}
