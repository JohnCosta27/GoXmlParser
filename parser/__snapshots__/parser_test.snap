
[TestSimple - 1]
Tag
 OpenTag
 Name: a
 ElementSuffixOpen
  ContentData
  Data: hello
  CloseTag
  Name: a

---

[TestNesting - 1]
Tag
 OpenTag
 Name: a
 ElementSuffixOpen
  ContentElement
   Tag
    OpenTag
    Name: b
    ElementSuffixOpen
     ContentData
     Data: Something
     CloseTag
     Name: b
  CloseTag
  Name: a

---

[TestSelfClosing - 1]
Tag
 OpenTag
 Name: atag
  Attributes
  Name: bruh
  String: hello-world
 ElementSuffixClose

---

[TestAttributes - 1]
Tag
 OpenTag
 Name: a
  Attributes
  Name: hello
  String: world
 ElementSuffixOpen
  CloseTag
  Name: a

---

[TestBigObject - 1]
Tag
 OpenTag
 Name: Tests
  Attributes
  Name: xmlns
  String: http://www.adatum.com
 ElementSuffixOpen
  ContentElement
   Tag
    OpenTag
    Name: Test
     Attributes
     Name: TestId
     String: 0001
      Attributes
      Name: TestType
      String: CMD
    ElementSuffixOpen
     ContentElement
      Tag
       OpenTag
       Name: Name
       ElementSuffixOpen
        ContentData
        Data: Convert number to string
        CloseTag
        Name: Name
      ContentElement
       Tag
        OpenTag
        Name: CommandLine
        ElementSuffixOpen
         ContentData
         Data: Examp1.EXE
         CloseTag
         Name: CommandLine
       ContentElement
        Tag
         OpenTag
         Name: Input
         ElementSuffixOpen
          ContentData
          Data: 1
          CloseTag
          Name: Input
        ContentElement
         Tag
          OpenTag
          Name: Output
          ElementSuffixOpen
           ContentData
           Data: One
           CloseTag
           Name: Output
     CloseTag
     Name: Test
   ContentElement
    Tag
     OpenTag
     Name: Test
      Attributes
      Name: TestId
      String: 0002
       Attributes
       Name: TestType
       String: CMD
     ElementSuffixOpen
      ContentElement
       Tag
        OpenTag
        Name: Name
        ElementSuffixOpen
         ContentData
         Data: Find succeeding characters
         CloseTag
         Name: Name
       ContentElement
        Tag
         OpenTag
         Name: CommandLine
         ElementSuffixOpen
          ContentData
          Data: Examp2.EXE
          CloseTag
          Name: CommandLine
        ContentElement
         Tag
          OpenTag
          Name: Input
          ElementSuffixOpen
           ContentData
           Data: abc
           CloseTag
           Name: Input
         ContentElement
          Tag
           OpenTag
           Name: Output
           ElementSuffixOpen
            ContentData
            Data: def
            CloseTag
            Name: Output
      CloseTag
      Name: Test
    ContentElement
     Tag
      OpenTag
      Name: Test
       Attributes
       Name: TestId
       String: 0003
        Attributes
        Name: TestType
        String: GUI
      ElementSuffixOpen
       ContentElement
        Tag
         OpenTag
         Name: Name
         ElementSuffixOpen
          ContentData
          Data: Convert multiple numbers to strings
          CloseTag
          Name: Name
        ContentElement
         Tag
          OpenTag
          Name: CommandLine
          ElementSuffixOpen
           ContentData
           Data: Examp2.EXE /Verbose
           CloseTag
           Name: CommandLine
         ContentElement
          Tag
           OpenTag
           Name: Input
           ElementSuffixOpen
            ContentData
            Data: 123
            CloseTag
            Name: Input
          ContentElement
           Tag
            OpenTag
            Name: Output
            ElementSuffixOpen
             ContentData
             Data: One Two Three
             CloseTag
             Name: Output
       CloseTag
       Name: Test
     ContentElement
      Tag
       OpenTag
       Name: Test
        Attributes
        Name: TestId
        String: 0004
         Attributes
         Name: TestType
         String: GUI
       ElementSuffixOpen
        ContentElement
         Tag
          OpenTag
          Name: Name
          ElementSuffixOpen
           ContentData
           Data: Find correlated key
           CloseTag
           Name: Name
         ContentElement
          Tag
           OpenTag
           Name: CommandLine
           ElementSuffixOpen
            ContentData
            Data: Examp3.EXE
            CloseTag
            Name: CommandLine
          ContentElement
           Tag
            OpenTag
            Name: Input
            ElementSuffixOpen
             ContentData
             Data: a1
             CloseTag
             Name: Input
           ContentElement
            Tag
             OpenTag
             Name: Output
             ElementSuffixOpen
              ContentData
              Data: b1
              CloseTag
              Name: Output
        CloseTag
        Name: Test
      ContentElement
       Tag
        OpenTag
        Name: Test
         Attributes
         Name: TestId
         String: 0005
          Attributes
          Name: TestType
          String: GUI
        ElementSuffixOpen
         ContentElement
          Tag
           OpenTag
           Name: Name
           ElementSuffixOpen
            ContentData
            Data: Count characters
            CloseTag
            Name: Name
          ContentElement
           Tag
            OpenTag
            Name: CommandLine
            ElementSuffixOpen
             ContentData
             Data: FinalExamp.EXE
             CloseTag
             Name: CommandLine
           ContentElement
            Tag
             OpenTag
             Name: Input
             ElementSuffixOpen
              ContentData
              Data: This is a test
              CloseTag
              Name: Input
            ContentElement
             Tag
              OpenTag
              Name: Output
              ElementSuffixOpen
               ContentData
               Data: 14
               CloseTag
               Name: Output
         CloseTag
         Name: Test
       ContentElement
        Tag
         OpenTag
         Name: Test
          Attributes
          Name: TestId
          String: 0006
           Attributes
           Name: TestType
           String: GUI
         ElementSuffixOpen
          ContentElement
           Tag
            OpenTag
            Name: Name
            ElementSuffixOpen
             ContentData
             Data: Another Test
             CloseTag
             Name: Name
           ContentElement
            Tag
             OpenTag
             Name: CommandLine
             ElementSuffixOpen
              ContentData
              Data: Examp2.EXE
              CloseTag
              Name: CommandLine
            ContentElement
             Tag
              OpenTag
              Name: Input
              ElementSuffixOpen
               ContentData
               Data: Test Input
               CloseTag
               Name: Input
             ContentElement
              Tag
               OpenTag
               Name: Output
               ElementSuffixOpen
                ContentData
                Data: 10
                CloseTag
                Name: Output
          CloseTag
          Name: Test
  CloseTag
  Name: Tests

---

[TestMedium - 1]
Tag
 OpenTag
 Name: a
 ElementSuffixOpen
  ContentElement
   Tag
    OpenTag
    Name: b
    ElementSuffixOpen
     ContentData
     Data: hello
     CloseTag
     Name: b
   ContentElement
    Tag
     OpenTag
     Name: c
     ElementSuffixOpen
      ContentElement
       Tag
        OpenTag
        Name: d
        ElementSuffixOpen
         ContentData
         Data: world
         CloseTag
         Name: d
       ContentElement
        Tag
         OpenTag
         Name: e
         ElementSuffixOpen
          ContentData
          Data: world2
          CloseTag
          Name: e
        ContentElement
         Tag
          OpenTag
          Name: f
          ElementSuffixOpen
           ContentData
           Data: world3
           CloseTag
           Name: f
      CloseTag
      Name: c
  CloseTag
  Name: a

---

[TestDoubleTesting - 1]
Tag
 OpenTag
 Name: a
 ElementSuffixOpen
  ContentElement
   Tag
    OpenTag
    Name: c
    ElementSuffixOpen
     ContentData
     Data: Hello
     CloseTag
     Name: c
   ContentElement
    Tag
     OpenTag
     Name: b
     ElementSuffixOpen
      ContentElement
       Tag
        OpenTag
        Name: d
        ElementSuffixOpen
         ContentData
         Data: 123
         CloseTag
         Name: d
       ContentElement
        Tag
         OpenTag
         Name: d
         ElementSuffixOpen
          ContentData
          Data: 321
          CloseTag
          Name: d
      CloseTag
      Name: b
  CloseTag
  Name: a

---
